lib/instana/frameworks/roda.rb

```diff
@@ -1,6 +1,47 @@
 require "instana/rack"
 
+module Instana
+  module RodaPathTemplateExtractor
+    module RequestMethods
+      TERM = defined?(::Roda) ? ::Roda::RodaPlugins::Base::RequestMethods::TERM : Object
+      
+      def if_match(args, &blk)
+        path = @remaining_path
+        captures = @captures.clear
+  
+        if match_all(args)
+          (env['INSTANA_PATH_TEMPLATE_FRAGMENTS'] ||= []).concat(named_args(args, blk))
```

Comment:
This is a tricky part and I'm not sure I completely follow it. Why do we need to store the template fragments in an env variable?

Reply:
Roda (and Cuba) build a tree of matchers which is evaluated for each request. For large and complex trees, we won't see all of the matchers within the same call to `#if_match` This means that we need to store the successfully matched tokens somewhere that we can access it in subsequent recursive calls to `#if_match`. 

----------------------

To illustrate this, I added some extra logging locally to `#if_match` to show how we handle two Roda applications which generate the same output:

```ruby
class ExampleOne < Roda
  route do |r|
    r.on "greet" do
      r.get String do |name|
        "Hello, #{name}!"
      end
    end
  end
end
```

```ruby
class ExampleTwo < Roda
  route do |r|
    r.get "greet", String do |name|
        "Hello, #{name}!"
    end
  end
end
```

Both example applications will respond to `GET /greet/Stan` with `Hello, Stan!`. `ExampleOne` will require two calls to `#if_match` to generate the response. Its request logs look like

```
[2021-01-06 11:25:30] INFO  WEBrick 1.6.0
[2021-01-06 11:25:30] INFO  ruby 2.7.2 (2020-10-01) [x86_64-darwin20]
[2021-01-06 11:25:30] INFO  WEBrick::HTTPServer#start: pid=93162 port=9292
["Seen Args", ["greet"]]
["Seen Args", [String, TERM]]
["Path Template", "/greet/{name}"]
::1 - - [06/Jan/2021:11:25:32 -0500] "GET /greet/Stan HTTP/1.1" 200 12 0.0014
```

`ExampleTwo` will require one call to `#if_match` to generate the response. Its request logs look like:

```
[2021-01-06 11:26:39] INFO  WEBrick 1.6.0
[2021-01-06 11:26:39] INFO  ruby 2.7.2 (2020-10-01) [x86_64-darwin20]
[2021-01-06 11:26:39] INFO  WEBrick::HTTPServer#start: pid=93182 port=9292
["Seen Args", ["greet", String, TERM]]
["Path Template", "/greet/{name}"]
::1 - - [06/Jan/2021:11:26:41 -0500] "GET /greet/Stan HTTP/1.1" 200 12 0.0012
```

Reply:
I see, so the issue is with nested resources. Thanks for detailed explanation 👍